### PR TITLE
New windowing tracker and large refactor of IconTasklist

### PR DESCRIPTION
**Pull request by EbonJaeger**
## Description

This all started off as a replacement of Abomination with a new module leveraging `libxfce4windowing`. Because of the tight coupling between IconTasklist and Abomination, it quickly became way more than that, becoming a large-scale rewrite of IconTasklist.

A new module, named "Windowing," has been written as a way to wrap and extend `libxfce4windowing`, making code that uses it easier to maintain, as well as making more code reusable. It is a combination, in essence, of AppSys and Abomination.

The functionality of IconTasklist should be the same as the old version. If anything, there might be an additional feature or two. Previously, theme designers had no way to change the look of an IconButton if a window requested urgency. Now, a new CSS class (`needs-attention`) is applied when a window becomes urgent, and is removed when urgency is no longer needed.

Popovers for right-clicking an IconButton have seen a slight redesign. Pages for window controls are now created and destroyed on-demand, which should yield a small improvement to memory usage.

Some X11-specific code still exists; it's currently unavoidable while we're still an X11-only desktop environment. However, this puts us in a much better place to be able to support both X11 and Wayland in IconTasklist, and any other code that will use the new Windowing module.

### Submitter Checklist

- [ ] Squashed commits with `git rebase -i` (if needed)
- [x] Built budgie-desktop and verified that the patch worked (if needed)

### CODE REVIEW

File: src/panel/applets/icon-tasklist/Widgets/IconButton.vala [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+
+	public override bool scroll_event(Gdk.EventScroll event) {
+		if (get_monotonic_time() - last_scroll_time < SCROLL_TIMEOUT) {
+			return Gdk.EVENT_STOP;
+		}
+
+		// Nothing to do if there are no open windows
+		if (window_group == null) {
+			return Gdk.EVENT_STOP;
+		}
```
Comment (JoshStrobl) [inline src/panel/applets/icon-tasklist/Widgets/IconButton.vala:165]:
```suggestion
		if (window_group == null) return Gdk.EVENT_STOP;
```

File: src/panel/applets/icon-tasklist/Widgets/IconButton.vala [hunk 2]
```diff
@@ -0,0 +1,676 @@
...
+		if (window_group != null && window_group.has_windows()) {
+			windows = window_group.get_windows();
+		} else {
+			windows = new List<unowned libxfce4windowing.Window>();
+		}
+
+		// No indicators if there are no windows
+		if (windows.is_empty()) {
+			return base.draw(ctx);
+		}
```
Comment (JoshStrobl) [inline src/panel/applets/icon-tasklist/Widgets/IconButton.vala:237]:
```suggestion
		if (windows.is_empty()) return base.draw(ctx);
```

File: src/panel/applets/icon-tasklist/Widgets/IconButton.vala [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+
+		// If this button does not have any focused windows,
+		// draw the inactive versions of the window indicators
+		if (!get_active()) {
+			return draw_inactive(ctx);
+		}
```
Comment (JoshStrobl) [inline src/panel/applets/icon-tasklist/Widgets/IconButton.vala:243]:
```suggestion
		if (!get_active()) return draw_inactive(ctx);
```

File: src/panel/applets/icon-tasklist/Widgets/IconButton.vala [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+		return window_group;
+	}
+
+	public void set_urgent(bool urgent) {
+		this.urgent = urgent;
+
+		if (urgent) {
+			get_style_context().add_class("needs-attention");
+			// TODO: Can this be left to themes?
+			icon.animate_attention(panel_position);
```
Comment (JoshStrobl) [inline src/panel/applets/icon-tasklist/Widgets/IconButton.vala:563]: What about the internal theme?

File: src/panel/applets/icon-tasklist/Widgets/IconButton.vala [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+			get_style_context().remove_class("running");
+
+			if (pinned) {
+				window_group = null;
+			} else {
+				return;
+			}
+
+			var active_window = window_group?.get_active_window() ?? window_group?.get_last_active_window();
+			set_tooltip_text(app?.name ?? active_window?.get_name() ?? "");
```
Comment (JoshStrobl) [inline src/panel/applets/icon-tasklist/Widgets/IconButton.vala:640]:
```suggestion
			if (!pinned) return;
			window_group = null;

			var active_window = window_group?.get_active_window() ?? window_group?.get_last_active_window();
			set_tooltip_text(app?.name ?? active_window?.get_name() ?? "");
```

File: src/panel/applets/icon-tasklist/Widgets/IconButton.vala [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+		} else {
+			icon.set_from_icon_name("image-missing", Gtk.IconSize.INVALID);
+		}
+
+		if (target_icon_size > 0) {
+			icon.pixel_size = target_icon_size;
+		} else {
+			// prevents apps making the panel massive when the icon initially gets added
+			icon.pixel_size = DEFAULT_ICON_SIZE;
+		}
```
Comment (JoshStrobl) [inline src/panel/applets/icon-tasklist/Widgets/IconButton.vala:674]:
```suggestion
		// prevents apps making the panel massive when the icon initially gets added
		icon.pixel_size = (target_icon_size > 0) ? target_icon_size : DEFAULT_ICON_SIZE;
```

File: src/windowing/window_mapper.vala [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+
+			// Try to get the application based on GtkApplication ID
+			var gtk_id = query_atom_string((ulong) window.x11_get_xid(), Gdk.Atom.intern("_GTK_APPLICATION_ID", false), true);
+
+			if (gtk_id != null) {
+				var desktop_id = gtk_id.down() + ".desktop";
+
+				if (desktop_id in applications) {
+					return applications[desktop_id];
+				}
```
Comment (JoshStrobl) [inline src/windowing/window_mapper.vala:303]:
```suggestion
				if (desktop_id in applications) return applications[desktop_id];
```

File: src/windowing/window_mapper.vala [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+				class_group_name = get_group_name(window);
+			}
+
+			// Try to match the class group name to an application
+			if (class_group_name != null) {
+				var info = query_name(class_group_name);
+
+				if (info != null) {
+					return info;
+				}
```
Comment (JoshStrobl) [inline src/windowing/window_mapper.vala:319]:
```suggestion
				if (info != null) return info;
```

File: src/windowing/window_mapper.vala [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+
+			// Try to match the class instance name to an application
+			unowned var instance_name = wnck_window.get_class_instance_name();
+
+			if (instance_name != null) {
+				var info = query_name(instance_name);
+
+				if (info != null) {
+					return info;
+				}
```
Comment (JoshStrobl) [inline src/windowing/window_mapper.vala:330]:
```suggestion
				if (info != null) return info;
```

**Review by JoshStrobl — CHANGES_REQUESTED** [review]

File: src/panel/applets/icon-tasklist/Widgets/ButtonPopover.vala [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+		construct set {
+			_pinned = value;
+
+			if (pin_button == null) return;
+
+			if (_pinned) {
+				pin_button.image = new Gtk.Image.from_icon_name("emblem-favorite", Gtk.IconSize.SMALL_TOOLBAR);
+			} else {
+				pin_button.image = new Gtk.Image.from_icon_name("emblem-favorite-symbolic", Gtk.IconSize.SMALL_TOOLBAR);
+			}
```
[the remark this answers is not in the document]
Comment (JoshStrobl) [reply src/panel/applets/icon-tasklist/Widgets/ButtonPopover.vala:36]: This code can go in, after convo with Evan I'm gonna implement a custom heart icon and one with a green accent / fill for use when it's favorited.

The goal of this change was to have it stand out when favorited and not use the star, and under Fluent it looks good but Papirus it looks like hot garbage. So easiest is doing it ourselves.

**Review by JoshStrobl — APPROVED** [review]

File: data/icons/status/budgie-app-unfavorited.svg
Comment (serebit) [inline data/icons/status/budgie-app-unfavorited.svg:1]: Could these be themeable? I'm iffy on making these static green and dark grey - depending on the GTK theme, the unfavorited state could be really hard to see. Arc-Dark has this problem (yes I know it's unsupported), and I'm sure others do as well. Maybe a heart outline for unfavorited and a filled in heart with a themeable color for favorited? Similar idea to the way we handle the notifications applet.

File: src/theme/common/_colors.scss
Comment (serebit) [inline src/theme/common/_colors.scss:1]: Por que?

File: src/windowing/window_mapper.vala [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+				name = "google-chrome";
+			}
+
+			return name;
+		}
+
+		private string? query_atom_string(ulong xid, Gdk.Atom atom, bool utf8) {
+			uint8[]? data = null;
+			Gdk.Atom a_type;
+			int a_f; // TODO: wat?
```
Comment (serebit) [inline src/windowing/window_mapper.vala:176]: wat?

**Review by serebit — CHANGES_REQUESTED** [review]
Code mostly looks good, but I do have some items for clarification and one that I think we should change.

File: src/theme/common/_colors.scss
Comment (JoshStrobl) [reply src/theme/common/_colors.scss:1]: Others were hex, makes it easier to copy into places like Inkscape.

**Review by serebit — APPROVED** [review]

File: data/icons/status/budgie-app-unfavorited.svg
Comment (serebit) [reply data/icons/status/budgie-app-unfavorited.svg:1]: Resolving conversation—this will be fixed separately.

**Pull request merged**